In [1]:
# Importing data tools
import pandas as pd

# Importing data visual tools
import numpy as np
import matplotlib.pyplot as plt

# Importing Scaler
from sklearn.preprocessing import StandardScaler

# Importing Sklearn Pipeline
from sklearn.pipeline import Pipeline

# Importing Sklearn Models
from sklearn.ensemble import AdaBoostClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm

from sklearn.ensemble import (
    AdaBoostClassifier, BaggingClassifier, ExtraTreesClassifier, 
    GradientBoostingClassifier, RandomForestClassifier
)
from sklearn.linear_model import (
    LogisticRegression, PassiveAggressiveClassifier, Perceptron, 
    RidgeClassifier, RidgeClassifierCV, SGDClassifier
)

from sklearn.svm import LinearSVC, NuSVC, SVC
from sklearn.naive_bayes import BernoulliNB, GaussianNB
from sklearn.neighbors import KNeighborsClassifier, NearestCentroid
from sklearn.tree import DecisionTreeClassifier, ExtraTreeClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.semi_supervised import LabelPropagation, LabelSpreading
from sklearn.dummy import DummyClassifier
from sklearn.calibration import CalibratedClassifierCV

# Importing sklearn metrics
from sklearn.metrics import (
    roc_auc_score, average_precision_score, precision_recall_curve, roc_curve, auc,
    f1_score, precision_score, recall_score, confusion_matrix, brier_score_loss, accuracy_score
)

# Importing sklearn clustering
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA, TruncatedSVD

# Importing Sklearn Samplers
from sklearn.model_selection import StratifiedKFold, StratifiedShuffleSplit, train_test_split

# Importing Imbalance-Learn 
from imblearn.pipeline import Pipeline as ImbPipeline
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler
from imblearn.combine import SMOTETomek

# Importing time
import time

In [2]:
n_split = 5
random_state = 28

# 1. SIMPLE E.D.A

In [3]:
X = pd.read_csv("X_creditcard.csv")
y = pd.read_csv("y_creditcard.csv")

In [4]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 35 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   V1                     284807 non-null  float64
 1   V2                     284807 non-null  float64
 2   V3                     284807 non-null  float64
 3   V4                     284807 non-null  float64
 4   V5                     284807 non-null  float64
 5   V6                     284807 non-null  float64
 6   V7                     284807 non-null  float64
 7   V8                     284807 non-null  float64
 8   V9                     284807 non-null  float64
 9   V10                    284807 non-null  float64
 10  V11                    284807 non-null  float64
 11  V12                    284807 non-null  float64
 12  V13                    284807 non-null  float64
 13  V14                    284807 non-null  float64
 14  V15                    284807 non-nu

# 2. APPLYING CLUSTERING TECHNIQUES

## We can apply clustring technique with 2 components (Fraud and Not Fraud). But it takes nearly 1hr and still the clustering has not been finished due to size of our dataset. So I decided to cancel our kernel.

In [ ]:
# # T-SNE
# t0 = time.time()
# X_reduced_tsne = TSNE(n_components=2, random_state=random_state).fit_transform(X.to_numpy())
# t1 = time.time()
# print("T-SNE took {:.2f} s".format(t1 - t0))

# # PCA
# t0 = time.time()
# X_reduced_pca = PCA(n_components=2, svd_solver='randomized', random_state=random_state).fit_transform(X.to_numpy())
# t1 = time.time()
# print("PCA took {:.2f} s".format(t1 - t0))

# # Truncated SVD
# t0 = time.time()
# X_reduced_svd = TruncatedSVD(n_components=2, algorithm='randomized', random_state=random_state).fit_transform(X.to_numpy())
# t1 = time.time()
# print("Truncated SVD took {:.2f} s".format(t1 - t0))

In [ ]:
# fig, axes = plt.subplots(1, 3, figsize=(18, 5))

# # T-SNE
# axes[0].scatter(X_reduced_tsne[:, 0], X_reduced_tsne[:, 1], c=y, cmap='tab10', s=10)
# axes[0].set_title("T-SNE (2D)")

# # PCA
# axes[1].scatter(X_reduced_pca[:, 0], X_reduced_pca[:, 1], c=y, cmap='tab10', s=10)
# axes[1].set_title("PCA (2D)")

# # Truncated SVD
# axes[2].scatter(X_reduced_svd[:, 0], X_reduced_svd[:, 1], c=y, cmap='tab10', s=10)
# axes[2].set_title("Truncated SVD (2D)")

# plt.suptitle("Dimensionality Reduction Comparison", fontsize=16)
# plt.tight_layout()
# plt.show()

# 3. KFOLD RANDOM SPLITTING (TRAIN / TEST)

In [14]:
sss = StratifiedShuffleSplit(n_splits=n_split, test_size=0.2, random_state=random_state)
stkf = StratifiedKFold(n_splits=n_split, shuffle=True, random_state=random_state)

for train_index, test_index in sss.split(X, y):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]
print(f'Stratfied Shuffle Split: shape of X_train is {X_train.shape}, shape of X_test is {X_test.shape}, shape of y_train is {y_train.shape}, shape of y_test is {y_test.shape}')

for train_index, test_index in stkf.split(X, y):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]
print(f'Stratfied KFold Split: shape of X_train is {X_train.shape}, shape of X_test is {X_test.shape}, shape of y_train is {y_train.shape}, shape of y_test is {y_test.shape}')

Stratfied Shuffle Split: shape of X_train is (227845, 35), shape of X_test is (56962, 35), shape of y_train is (227845, 1), shape of y_test is (56962, 1)
Stratfied KFold Split: shape of X_train is (227846, 35), shape of X_test is (56961, 35), shape of y_train is (227846, 1), shape of y_test is (56961, 1)


# 4. UNDERSAMPLING TECHNIQUES FOR IMBALANCED DATASET

## Here are the 3 of the most common UnderSampling Methods when dealing with Imbalanced Dataset

1. RandomUnderSampler - Randomly removes samples from the majority class until balance is achieved.
2. SMOTE (Synthetic Minority Oversampling Technique) - Generates synthetic samples for the minority class by interpolating between neighbors.
3. SMOTE‑Tomek - Combines SMOTE oversampling with Tomek links (removes borderline samples where classes overlap).

## Practical Applications
- If your dataset is large → start with RandomUnderSampler (quick, efficient).
- If your dataset is small or minority class is tiny → use SMOTE to generate synthetic samples.
- If your dataset has class overlap/noisy boundaries → use SMOTE‑Tomek for both balancing and cleaning.

On this dataset which is large - 284 thousand rows, we'll use SMOTE-Tomek Undersampling for both balanced result

# 5. TRAINING MODEL WITHOUT UNDERSAMPLING

In [4]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,        # 20% test, 80% train
    stratify=y,           # preserve class ratios
    random_state=random_state
)

X_train = X_train.to_numpy()
y_train = y_train.to_numpy().flatten()
X_test = X_test.to_numpy()
y_test = y_test.to_numpy().ravel()

In [5]:
models = {
    "LogisiticRegression": LogisticRegression(),
    "KNearest": KNeighborsClassifier(),
    "Support Vector Classifier": SVC(),
    "DecisionTreeClassifier": DecisionTreeClassifier(),
    "AdaBoostClassifier": AdaBoostClassifier()
}

In [6]:
results = {
    "Model": [],
    "Accuracy": [],
    "F1 Score Weighted": [],
    "F1 Score Macro": [],
    "Precision_weighted": [],
    "Recall_weighted": [],
    "Precision_macro_avg": [],
    "Recall_macro_avg": []
}

for model_name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    
    results["Model"].append(model_name)
    results["Accuracy"].append(accuracy_score(y_test, y_pred))
    results["F1 Score Weighted"].append(f1_score(y_test, y_pred, average='weighted'))
    results["F1 Score Macro"].append(f1_score(y_test, y_pred, average='macro'))
    results["Precision_weighted"].append(precision_score(y_test, y_pred, average='weighted'))
    results["Recall_weighted"].append(recall_score(y_test, y_pred, average='weighted'))
    results["Precision_macro_avg"].append(precision_score(y_test, y_pred, average='macro'))
    results["Recall_macro_avg"].append(recall_score(y_test, y_pred, average='macro'))

### RESULT W/OUT UNDERSAMPLING

In [9]:
results_df = pd.DataFrame(results)
print(results_df)

                       Model  Accuracy  F1 Score Weighted  F1 Score Macro  \
0        LogisiticRegression  0.999175           0.999103        0.859075   
1                   KNearest  0.999386           0.999345        0.898690   
2  Support Vector Classifier  0.999421           0.999393        0.907676   
3     DecisionTreeClassifier  0.999034           0.999056        0.865612   
4         AdaBoostClassifier  0.999140           0.999129        0.871512   

   Precision_weighted  Recall_weighted  Precision_macro_avg  Recall_macro_avg  
0            0.999109         0.999175             0.934449          0.806043  
1            0.999353         0.999386             0.959745          0.851988  
2            0.999391         0.999421             0.950398          0.872379  
3            0.999082         0.999034             0.850265          0.882372  
4            0.999119         0.999140             0.881483          0.862051  


# 6. TRAINING MODEL WITH UNDERSAMPLING

In [26]:
smote = SMOTETomek(random_state=random_state)

X_resampled, y_resampled = smote.fit_resample(X_train, y_train)

print("X_train Before SMOTETomek: \n", pd.Series(y_train).value_counts())
print("X_train After SMOTETomek: \n", pd.Series(y_resampled).value_counts())

print("y_train Before SMOTETomek: \n", pd.Series(y_train).value_counts())
print("y_train After SMOTETomek: \n", pd.Series(y_resampled).value_counts())

X_train Before SMOTETomek: 
 0    227451
1       394
Name: count, dtype: int64
X_train After SMOTETomek: 
 0    227451
1    227451
Name: count, dtype: int64
y_train Before SMOTETomek: 
 0    227451
1       394
Name: count, dtype: int64
y_train After SMOTETomek: 
 0    227451
1    227451
Name: count, dtype: int64


In [23]:
undersampled_models = {
    "LogisiticRegression": LogisticRegression(),
    "KNearest": KNeighborsClassifier(),
    "Support Vector Classifier": SVC(),
    "DecisionTreeClassifier": DecisionTreeClassifier(),
    "AdaBoostClassifier": AdaBoostClassifier()
}

In [24]:
undersampled_results = {
    "Model": [],
    "Accuracy": [],
    "F1 Score Weighted": [],
    "F1 Score Macro": [],
    "Precision_weighted": [],
    "Recall_weighted": [],
    "Precision_macro_avg": [],
    "Recall_macro_avg": []
}

# Train and evaluate
for model_name, model in undersampled_models.items():
    model.fit(X_resampled, y_resampled)
    y_pred = model.predict(X_test)
    
    undersampled_results["Model"].append(model_name)
    undersampled_results["Accuracy"].append(accuracy_score(y_test, y_pred))
    undersampled_results["F1 Score Weighted"].append(f1_score(y_test, y_pred, average='weighted'))
    undersampled_results["F1 Score Macro"].append(f1_score(y_test, y_pred, average='macro'))
    undersampled_results["Precision_weighted"].append(precision_score(y_test, y_pred, average='weighted'))
    undersampled_results["Recall_weighted"].append(recall_score(y_test, y_pred, average='weighted'))
    undersampled_results["Precision_macro_avg"].append(precision_score(y_test, y_pred, average='macro'))
    undersampled_results["Recall_macro_avg"].append(recall_score(y_test, y_pred, average='macro'))

C:\Users\kenne\miniconda3\envs\imblearn_py3.12\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### RESULT W/ UNDERSAMPLING

In [25]:
results_df = pd.DataFrame(undersampled_results)
print(results_df)

                       Model  Accuracy  F1 Score Weighted  F1 Score Macro  \
0        LogisiticRegression  0.973895           0.985242        0.547366   
1                   KNearest  0.997332           0.997815        0.752578   
2  Support Vector Classifier  0.985446           0.991238        0.579747   
3     DecisionTreeClassifier  0.997683           0.997987        0.751299   
4         AdaBoostClassifier  0.975054           0.985842        0.549334   

   Precision_weighted  Recall_weighted  Precision_macro_avg  Recall_macro_avg  
0            0.998234         0.973895             0.528608          0.946179  
1            0.998567         0.997332             0.685538          0.896799  
2            0.998172         0.985446             0.546132          0.916312  
3            0.998421         0.997683             0.699132          0.840949  
4            0.998220         0.975054             0.529566          0.941666  


# A. CONCLUSION
Comparing Results:

1. Classification without undersampling  
The metrics appeared extremely high (close to 100%). However, the model was biased toward the majority class. This can look “perfect,” but in reality minority detection was weak and the model risked overfitting.

3. Classification with undersampling applied (SMOTE‑Tomek)  
The metrics dropped (Macro F1 from ~0.86 down to ~0.54). This may seem “weaker,” but the model is now forced to treat both classes equally by balancing the minority class through undersampling. In domains such as fraud detection, this trade‑off is valuable, since correctly identifying minority cases is more important than inflated overall accuracy.

# B. VERDICT

We can use the undersampled K-Nearest Neighbors (KNN) classifier since based on the evaluation metrics, it achieved the highest performance among all models. 
Going back to data description. The dataset has already been transformed using PCA, which enhances separability and contributes to stronger results. KNN is particularly effective when working with clustered datasets, making it a suitable choice here.

Weighing accuracy vs F1 and precision/recall is critical.

It depends on your domain in acheiving business goal.

a. Accuracy - Good for overall performance, but can mislead if dataset is imbalanced (like our dataset).
b. F1 Score - Harmonic mean of precision and recall. It balances precision and recall in a single metric.
c. Precision - Out of all the samples the model predicted as positive, how many were actually positive? It avoids false alarms.
d. Recall - Out of all the actual positive samples, how many did the model correctly identify? Goal is to catch all frauds including possible non fraud transaction.

e. Weighted - Good when overall accuracy matters, but can hide minority class weakness.
f. Macro - Best for imbalanced datasets where minority detection is critical.

To further optimize performance and ensure the model is deployable, we can fine-tune its parameters using Optuna, enabling automated hyperparameter search and robust model selection.

We should also try to visualize our model performance.

1. Decision Boundary Plots
2. Feature Importance
3. Confusion Matrix
4. ROC Curve & AUC
5. Precision–Recall Curve
6. SHAP / LIME Explanations
7. PCA / SVD / T-SNE / Hierarchal Clustering / DBScan

# C. LIMITATION OF THIS STUDY

Please note that I did not include other classifiers (such as Random Forest, XGB, AdaBoost) in this time. 
Training five models already required about 45 minutes due to dataset size with 285k rows.
While Random Forest leverages multiple decision trees through bagging.
Given our computational resources and time constraints, we will focus on the five models already trained.

I did not also include other visualization technique due to same reason.

For future work, we can explore more precise models using AWS, Azure, GCP or other cloud-based IaaS/PaaS solutions that provide 
powerful computational resources.

# Please see the third part of this study - We'll use deep learning classification techniques to compare to the models mentioned above.